### ISSUE TITLE
Support mixed boxed and unboxed primitives in assertEquals() & assertNotEquals() 

### ISSUE THREAD

**Issue by ridono** [junit-team/junit5#1638]
## Overview

Asserting that an `int` equals an `Integer` is not possible with `assertEquals()` due to the compiler not knowing if it should use `assertEquals(int, int)` or `assertEquals(Object, Object)`.

This results in the compile error `The method assertEquals(int, int) is ambiguous for the type Assertions` like in the following example:

```java
	@Test
	void test_parseYear() throws Exception {
		Integer parsedYear = SeparateDateParser.parseYear("2014");
		Assertions.assertEquals(2014, parsedYear); // does not compile
	}
```

Instead you either have to declare another variable:

```java
	@Test
	void test_parseYear() throws Exception {
		Integer parsedYear = SeparateDateParser.parseYear("2014");
		Integer expectedYear = 2014;
		Assertions.assertEquals(expectedYear, parsedYear);
	}
```

or use a custom assertion method:

```java
	void assertEquals(Number expected, Number actual) {
		Assertions.assertEquals(expected, actual);
	}
```

We have already encapsulated this method in a utility class to reuse it in our projects, but as this is a quite common case it would be nice if the framework could provide an `assertEquals()` method like the one above.

## Related Issues

- #1607 

## Deliverables

- [ ] Support mixed boxed and unboxed primitives in `assertEquals()`
- [ ] Support mixed boxed and unboxed primitives in `assertNotEquals() `
- [ ] Document in Release Notes

**Comment by sbrannen** [comment]
I agree that that is a bit of a nuisance.

Note that the following is a workaround I often use:

```java
@Test
void test_parseYear() throws Exception {
	Integer parsedYear = Integer.valueOf(2014);
	Assertions.assertEquals(2014, parsedYear.intValue()); // does not compile
}
```

**Comment by sbrannen** [comment]
> I agree that that is a bit of a nuisance.

Coming back to that...

I suppose this nuisance exists not only for numbers but rather for all primitive types when assert equality between a primitive and its wrapper type.

So perhaps we should investigate a solution for all such cases.

@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
Tentatively slated for 5.4 M1 for _team discussion_.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
**Analysis**

Declaring methods like the one proposed in this issue's description only works if the methods are declared **locally** as in the following example.

```java
class MixedBoxedAndUnboxedPrimitivesTests {

	void assertEquals(Number expected, Number actual) {
		Assertions.assertEquals(expected, actual);
	}

	void assertEquals(Character expected, Character actual) {
		Assertions.assertEquals(expected, actual);
	}

	void assertEquals(Boolean expected, Boolean actual) {
		Assertions.assertEquals(expected, actual);
	}

	@Test
	void bytes() {
		assertEquals((byte) 42, Byte.valueOf("42"));
	}

	@Test
	void shorts() {
		assertEquals((short) 42, Short.valueOf("42"));
	}

	@Test
	void integers() {
		assertEquals(42, Integer.valueOf("42"));
	}

	@Test
	void longs() {
		assertEquals(42L, Long.valueOf("42"));
	}

	@Test
	void floats() {
		assertEquals(42.0f, Float.valueOf("42.0"));
	}

	@Test
	void doubles() {
		assertEquals(42.0d, Double.valueOf("42.0"));
	}

	@Test
	void booleans() {
		assertEquals(false, Boolean.valueOf("false"));
	}

	@Test
	void chars() {
		assertEquals('a', Character.valueOf('a'));
	}

}
```

Declaring those same methods as `static` in `Assertions` does _not_ remove the ambiguity for the compiler. 😞

**Comment by sbrannen** [comment]
This has been addressed in `master` in 0ec5a7bf10292efe1513e1f66cbab61a89d67537 and 382e9ea35cd5dfd9789ed28a0c9d16673cb5a485.
